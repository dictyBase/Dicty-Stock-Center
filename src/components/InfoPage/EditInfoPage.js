// @flow
import React, { Component } from "react"
import { connect } from "react-redux"
import { withStyles } from "@material-ui/core/styles"
import Grid from "@material-ui/core/Grid"
import Button from "@material-ui/core/Button"
import { EditorState, convertToRaw, convertFromRaw } from "draft-js"
import Editor from "draft-js-plugins-editor"
import createUndoPlugin from "draft-js-undo-plugin"
import createToolbarPlugin from "draft-js-static-toolbar-plugin"
import createToolbarLinkPlugin from "draft-js-toolbar-link-plugin"
import {
  ItalicButton,
  BoldButton,
  UnderlineButton,
  CodeButton,
  HeadlineOneButton,
  HeadlineTwoButton,
  HeadlineThreeButton,
  UnorderedListButton,
  OrderedListButton,
  BlockquoteButton,
  CodeBlockButton,
} from "draft-js-buttons"
import { saveEditing, cancelEditing } from "actions/page"
import { NAMESPACE } from "constants/dsctypes"
import "draft-js-static-toolbar-plugin/lib/plugin.css"

// Set up Draft.js toolbar and plugins
const undoPlugin = createUndoPlugin()
const toolbarLinkPlugin = createToolbarLinkPlugin({
  inputPlaceholder: "Insert URL here...",
})
const { LinkButton } = toolbarLinkPlugin
const { UndoButton, RedoButton } = undoPlugin
const toolbarPlugin = createToolbarPlugin({
  structure: [
    BoldButton,
    ItalicButton,
    UnderlineButton,
    CodeButton,
    HeadlineOneButton,
    HeadlineTwoButton,
    HeadlineThreeButton,
    UnorderedListButton,
    OrderedListButton,
    BlockquoteButton,
    CodeBlockButton,
    LinkButton,
    UndoButton,
    RedoButton,
  ],
})
const { Toolbar } = toolbarPlugin
const plugins = [toolbarPlugin, toolbarLinkPlugin, undoPlugin]

const styles = theme => ({
  editor: {
    boxSizing: "border-box",
    border: "1px solid #ddd",
    cursor: "text",
    padding: "10px",
    borderRadius: "2px",
    marginBottom: "2em",
    background: "#fefefe",
    margin: "10px auto",

    "& a": {
      color: "#004080",
      textDecoration: "none",
    },
  },
  container: {
    paddingRight: "15px",
    paddingLeft: "15px",
    marginRight: "auto",
    marginLeft: "auto",
    width: "90%",
    "@media (min-width: 1300px)": {
      width: "80%",
    },
  },
  toolbar: {
    backgroundColor: "#fafafa",
    borderRadius: "2px",
    border: "1px solid #ddd",
    padding: "5px",
    width: "100%",
    display: "inline-block",
  },
  buttonGrid: {
    marginRight: "3px",
  },
  saveButton: {
    width: "100%",
    backgroundColor: "#15317e",
  },
  cancelButton: {
    width: "100%",
  },
})

type Props = {
  /** React Router's match object */
  match: Object,
  /** action creator to cancel editing */
  cancelEditing: Function,
  /** the ID number for the page content */
  id: string,
  /** the user's ID number */
  userId: string,
  /** action creator to save page content */
  saveEditing: Function,
  /** Material-UI styling */
  classes: Object,
}

type State = {
  editorState: EditorState,
}

/**
 * Allows editing of the info page components (i.e. Deposit, Payment, Order)
 */

export class EditInfoPage extends Component<Props, State> {
  constructor(props: any) {
    super(props)

    if (props.content) {
      this.state = {
        editorState: EditorState.createWithContent(
          convertFromRaw(JSON.parse(props.content)),
        ),
      }
    }
  }

  onChange = (editorState: EditorState) => this.setState({ editorState })

  focus = () => this.refs.editor.focus()

  onSave = () => {
    const { editorState } = this.state
    const { id, saveEditing, userId } = this.props
    const rawData = JSON.stringify(
      convertToRaw(editorState.getCurrentContent()),
    )
    const body = {
      id,
      data: {
        id,
        type: "contents",
        attributes: {
          updated_by: userId,
          content: rawData,
        },
      },
    }
    saveEditing(id, body)
  }

  onCancel = () => {
    const { cancelEditing, match } = this.props
    cancelEditing(`${match.params.name}`)
  }

  render() {
    const { editorState } = this.state
    const { classes } = this.props

    return (
      <div className={classes.container}>
        <div className={classes.toolbar}>
          <Toolbar />
        </div>
        <div className={classes.editor}>
          <Editor
            editorState={editorState}
            onChange={this.onChange}
            plugins={plugins}
            ref="{(element) => { this.editor = element }}"
          />
        </div>
        <Grid container justify="space-between">
          <Grid item xs={3} />
          <Grid item xs={3} />
          <Grid item xs={2} className={classes.buttonGrid}>
            <Button
              className={classes.cancelButton}
              size="small"
              variant="contained"
              onClick={this.onCancel}>
              Cancel
            </Button>
          </Grid>
          <Grid item xs={2}>
            <Button
              className={classes.saveButton}
              size="small"
              variant="contained"
              color="primary"
              onClick={this.onSave}>
              Save
            </Button>
          </Grid>
        </Grid>
      </div>
    )
  }
}

const mapStateToProps = (state, ownProps) => {
  const slugName = `${NAMESPACE}-${ownProps.match.params.name}`
  return {
    content: state.page[slugName].data.attributes.content,
    id: state.page[slugName].data.id,
    userId: state.auth.user.data.id,
  }
}

export default connect<*, *, *, *, *, *>(
  mapStateToProps,
  { saveEditing, cancelEditing },
)(withStyles(styles)(EditInfoPage))
