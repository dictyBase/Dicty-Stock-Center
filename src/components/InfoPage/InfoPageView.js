/* eslint-disable react/jsx-no-bind */
// @flow
import React, { Component } from "react"
import { connect } from "react-redux"
import {
  Editor,
  EditorState,
  convertFromRaw,
  CompositeDecorator,
} from "draft-js"
import { withStyles } from "@material-ui/core/styles"
import findLinkEntities from "utils/findLinkEntities"
import { ContentAPI } from "utils/apiClasses"
import Link from "components/Link"
import Authorization from "components/authentication/Authorization"
import ErrorNotification from "components/authentication/ErrorNotification"
import timeSince from "utils/timeSince"
import { editPage } from "actions/page"
import { fetchUserInfo } from "actions/auth"
import { FontAwesomeIcon } from "@fortawesome/react-fontawesome"
import Grid from "@material-ui/core/Grid"
import { Container, ToolbarNav, TextInfo } from "styles"

const styles = theme => ({
  label: {
    display: "inline",
    padding: "0.2em 0.6em 0.3em",
    fontSize: "75%",
    fontWeight: "bold",
    lineHeight: 1,
    color: "#fff",
    textAlign: "center",
    whiteSpace: "nowrap",
    verticalAlign: "baseline",
    borderRadius: "0.25em",
    backgroundColor: "#337ab7",
    "&:hover": {
      backgroundColor: "#337ab7",
    },
    "&:focus": {
      backgroundColor: "#337ab7",
    },
  },
  inlineLink: {
    cursor: "pointer",
  },
  content: {
    marginLeft: "auto",
  },
})

const decorator = [
  {
    strategy: findLinkEntities,
    component: Link,
  },
]

const error =
  "Your login token is expired. Please log out and then log back in to regain full user access."

type Props = {
  /** React Router's match object */
  match: Object,
  /** action creator for editing the current page content */
  editPage: Function,
  /** action creator to fetch a non-authenticated user's information */
  fetchUserInfo: Function,
  /** the object that contains page data from current state */
  page: Object,
  /** boolean representing whether the user is logged in or not */
  isAuthenticated: boolean,
  /** Material-UI styling */
  classes: Object,
}

type State = {
  editorState: EditorState,
}

/** Displays the info page data that was fetched from the InfoPage component */

export class InfoPageView extends Component<Props, State> {
  constructor(props: any) {
    super(props)

    this.state = {
      editorState: EditorState.createWithContent(
        convertFromRaw(JSON.parse(props.page.data.attributes.content)),
        new CompositeDecorator(decorator),
      ),
    }
  }

  componentDidMount() {
    const { isAuthenticated, page, fetchUserInfo } = this.props
    if (isAuthenticated) {
      const fetchedUser = new ContentAPI(page).getUser()
      fetchUserInfo(fetchedUser)
    }
  }

  onChange = (editorState: EditorState) => this.setState({ editorState })

  onClick = (e: SyntheticEvent<>) => {
    e.preventDefault()

    const { editPage, match, page } = this.props
    editPage(page.data.attributes.content, match.params.name)
  }

  render() {
    const { updated_at } = this.props.page.data.attributes
    const { isAuthenticated, classes } = this.props

    return (
      <Container>
        {isAuthenticated && (
          //$FlowFixMe
          <Authorization
            render={({ canEditPages, fetchedUserData, verifiedToken }) => (
              <div>
                {canEditPages && verifiedToken === false && (
                  <ErrorNotification error={error} />
                )}
                <br />
                {canEditPages && fetchedUserData && (
                  <ToolbarNav>
                    <Grid container>
                      <Grid item>
                        <TextInfo>
                          <strong>
                            <FontAwesomeIcon icon="user" />{" "}
                            {fetchedUserData.getFullName()}
                          </strong>{" "}
                          edited {timeSince(updated_at)} ago
                        </TextInfo>
                      </Grid>
                      <Grid item className={classes.content}>
                        <span className={classes.label}>
                          {fetchedUserData.getRoles()}
                        </span>{" "}
                        &nbsp,
                        {verifiedToken && (
                          // eslint-disable-next-line jsx-a11y/anchor-is-valid
                          <a
                            className={classes.inlineLink}
                            onClick={this.onClick}>
                            <FontAwesomeIcon
                              icon="pencil-alt"
                              title="Edit page"
                            />
                          </a>
                        )}
                      </Grid>
                    </Grid>
                  </ToolbarNav>
                )}
              </div>
            )}
          />
        )}

        <Grid container>
          <Grid item>
            <Editor
              editorState={this.state.editorState}
              ref="editor"
              readOnly
              onChange={this.onChange}
              decorators={decorator}
            />
          </Grid>
        </Grid>
      </Container>
    )
  }
}

type authState = {
  auth: {
    isAuthenticated: boolean,
  },
}

export const mapStateToProps = (state: authState) => ({
  isAuthenticated: state.auth.isAuthenticated,
})

export default connect(
  mapStateToProps,
  { editPage, fetchUserInfo },
)(withStyles(styles)(InfoPageView))
