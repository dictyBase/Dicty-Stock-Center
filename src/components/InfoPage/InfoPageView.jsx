// @flow
import React, { Component } from "react"
import { connect } from "react-redux"
import {
  Editor,
  EditorState,
  convertFromRaw,
  CompositeDecorator
} from "draft-js"
import findLinkEntities from "utils/findLinkEntities"
import Link from "components/Link"
import timeSince from "utils/timeSince"
import { editPage } from "actions/page"
import FontAwesome from "react-fontawesome"
import { Flex, Box } from "rebass"
import { Container, ToolbarNav, TextInfo, Label, InlineLink } from "styles"

const decorator = [
  {
    strategy: findLinkEntities,
    component: Link
  }
]

type Props = {
  /** React Router's match object */
  match: Object,
  /** user who last updated the page */
  updated_by: string,
  /** time the page was last updated */
  updated_at: string,
  /** action creator for editing the current page content */
  editPage: Function,
  /** checks if user is authenticated */
  isAuthenticated: boolean,
  /** the object that contains page data from current state */
  page: Object
}

type State = {
  editorState: EditorState
}

/** Displays the info page data that was fetched from the InfoPage component */

class InfoPageView extends Component<Props, State> {
  constructor(props) {
    super(props)

    this.state = {
      editorState: EditorState.createWithContent(
        convertFromRaw(JSON.parse(props.page.content)),
        new CompositeDecorator(decorator)
      )
    }
  }
  onChange = editorState => this.setState({ editorState })
  onClick = e => {
    e.preventDefault()

    const { editPage, match, page } = this.props
    editPage(page.content, match.params.name)
  }
  render() {
    const { updated_by, updated_at } = this.props.page
    const { isAuthenticated, user } = this.props
    console.log(this.props)
    return (
      <Container>
        {isAuthenticated && (
          <ToolbarNav>
            <Flex>
              <Box>
                <TextInfo>
                  <strong>
                    <FontAwesome name="user" /> {updated_by}
                  </strong>{" "}
                  edited {timeSince(updated_at)} ago
                </TextInfo>
              </Box>
              <Box ml="auto">
                <div>
                  {/* need to implement custom label */}
                  <Label>Curator</Label> &nbsp; &nbsp;
                  <InlineLink onClick={this.onClick}>
                    <FontAwesome name="pencil" title="Edit page" />
                  </InlineLink>
                </div>
              </Box>
            </Flex>
          </ToolbarNav>
        )}

        <Flex>
          <Box>
            <Editor
              editorState={this.state.editorState}
              ref="editor"
              readOnly
              onChange={this.onChange}
              decorators={decorator}
            />
          </Box>
        </Flex>
      </Container>
    )
  }
}

const mapStateToProps = state => {
  return {
    isAuthenticated: state.auth.isAuthenticated,
    user: state.auth.user
  }
}

export default connect(mapStateToProps, { editPage })(InfoPageView)
