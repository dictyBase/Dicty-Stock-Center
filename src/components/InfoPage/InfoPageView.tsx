import React, { useRef, useState } from "react"
import { useHistory, useParams } from "react-router-dom"
import {
  Editor,
  EditorState,
  convertFromRaw,
  CompositeDecorator,
} from "draft-js"
import { withStyles } from "@material-ui/core/styles"
import Grid from "@material-ui/core/Grid"
import Tooltip from "@material-ui/core/Tooltip"
import IconButton from "@material-ui/core/IconButton"
import findLinkEntities from "utils/findLinkEntities"
import Link from "components/Link"
import ErrorNotification from "components/authentication/ErrorNotification"
import timeSince from "utils/timeSince"
import { FontAwesomeIcon } from "@fortawesome/react-fontawesome"
import { useAuthStore } from "components/authentication/AuthStore"
import useAuthorization from "hooks/useAuthorization"

const styles = () => ({
  container: {
    paddingRight: "15px",
    paddingLeft: "15px",
    marginRight: "auto",
    marginLeft: "auto",
    width: "90%",
    "@media (min-width: 1300px)": {
      width: "80%",
    },
  },
  label: {
    display: "inline",
    padding: "0.2em 0.6em 0.3em",
    fontSize: "75%",
    // fontWeight: "bold",
    lineHeight: 1,
    color: "#fff",
    // textAlign: "center",
    // whiteSpace: "nowrap",
    verticalAlign: "baseline",
    borderRadius: "0.25em",
    backgroundColor: "#337ab7",
    "&:hover": {
      backgroundColor: "#337ab7",
    },
    "&:focus": {
      backgroundColor: "#337ab7",
    },
  },
  inlineLink: {
    cursor: "pointer",
  },
  content: {
    marginLeft: "auto",
  },
  toolbar: {
    backgroundColor: "#fafafa",
    borderRadius: "2px",
    border: "1px solid #ddd",
    padding: "5px",
    width: "100%",
    display: "inline-block",
  },
  editButton: {
    fontSize: "0.99em",
    color: "#337ab7",
    "&:hover": {
      color: "#337ab7",
      backgroundColor: "transparent",
    },
  },
  textInfo: {
    color: "#31708f",
    "&:hover": {
      color: "#245269",
    },
    "&:focus": {
      color: "#245269",
    },
  },
  editor: {
    "& a": {
      color: "#004080",
      textDecoration: "none",
    },
  },
})

const decorator = [
  {
    strategy: findLinkEntities,
    component: Link,
  },
]

const error =
  "Your login token is expired. Please log out and then log back in to regain full user access."

type Props = {
  data: {
    content: string
    created_by: {
      id: string
      first_name: string
      last_name: string
      roles: Array<{
        role
      }>
    }
    updated_by: {
      id: string
      first_name: string
      last_name: string
      updated_at: string
      roles: Array<{
        role
      }>
    }
  }
  classes: {
    container: string
    toolbar: string
    textInfo: string
    content: string
    label: string
    editButton: string
    editor: string
  }
}

/** Displays the info page data that was fetched from the InfoPage component */

const InfoPageView = ({ data, classes }: Props) => {
  const [editorState, setEditorState] = useState(() =>
    EditorState.createWithContent(
      convertFromRaw(JSON.parse(data.content)),
      // @ts-ignore
      new CompositeDecorator(decorator),
    ),
  )
  const history = useHistory()
  const { name } = useParams()
  const [{ isAuthenticated }] = useAuthStore()
  const { canEditPages, verifiedToken } = useAuthorization()

  const editorRef = useRef<any>(null)
  const onEdit = event => {
    event.preventDefault()
    history.push({
      pathname: `/information/${name}/edit`,
      state: {
        data: data,
      },
    })
  }

  const fullName = `${data.updated_by.first_name} ${data.updated_by.last_name}`
  const role = `${data.updated_by.roles[0].role}`
  return (
    <div className={classes.container}>
      {isAuthenticated && canEditPages && verifiedToken && (
        <div>
          {canEditPages && !verifiedToken && (
            <ErrorNotification error={error} />
          )}
          <br />
          {canEditPages && (
            <div className={classes.toolbar}>
              <Grid container alignItems="center">
                <Grid item>
                  <span className={classes.textInfo}>
                    <strong>
                      <FontAwesomeIcon icon="user" /> {fullName}
                    </strong>{" "}
                    edited {timeSince(data.updated_by.updated_at)} ago
                  </span>
                </Grid>
                <Grid item className={classes.content}>
                  <span className={classes.label}>{role}</span> &nbsp;
                  {verifiedToken && (
                    <Tooltip title="Edit Page" placement="bottom">
                      <IconButton
                        className={classes.editButton}
                        onClick={onEdit}>
                        <FontAwesomeIcon icon="pencil-alt" />
                      </IconButton>
                    </Tooltip>
                  )}
                </Grid>
              </Grid>
            </div>
          )}
        </div>
      )}

      <Grid container>
        <Grid item className={classes.editor}>
          <Editor
            editorState={editorState}
            ref={editor => (editorRef.current = editor)}
            readOnly
            onChange={setEditorState}
            // @ts-ignore
            decorators={decorator}
          />
        </Grid>
      </Grid>
    </div>
  )
}

export default withStyles(styles)(InfoPageView)
