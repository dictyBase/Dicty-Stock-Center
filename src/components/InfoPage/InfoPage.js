// @flow
import React, { Component, Fragment } from "react"
import { connect } from "react-redux"
import { Helmet } from "react-helmet"
import Loader from "components/common/Loader"
import InfoPageView from "./InfoPageView"
import ErrorPage from "components/Errors/ErrorPage"
import { fetchInfoPage } from "actions/page"
import { NAMESPACE } from "constants/dsctypes"

type Props = {
  /** Checks if data is currently being fetched */
  isFetching: boolean,
  /** the Page object taken from the current state  */
  page: Object,
  /** React Router's match object */
  match: Object,
  /** Action creator that fetches data from API */
  fetchInfoPage: Function,
}

/**
 * Fetches the data for the desired info page
 */

class InfoPage extends Component<Props> {
  // set defaultProps to prevent console warnings
  static defaultProps = {
    page: {
      data: {
        attributes: {},
      },
    },
  }
  componentDidMount() {
    const { match, fetchInfoPage } = this.props
    const slugName = `${NAMESPACE}-${match.params.name}`
    fetchInfoPage(slugName)
  }
  render() {
    const { isFetching, page } = this.props
    const name = page.data.attributes.name

    if (!isFetching && page.error) {
      return <ErrorPage error={page.error} />
    }

    if (!isFetching && page.data.attributes.content) {
      return (
        <Fragment>
          <Helmet>
            <title>
              {name.charAt(0).toUpperCase() + name.slice(1)} Information - Dicty
              Stock Center
            </title>
            <meta
              name="description"
              content="The Dicty Stock Center is a rapidly growing central repository for Dictyostelium discoideum strains and those of related species, plasmids, commonly used food bacteria, and other materials such as antibodies."
            />
          </Helmet>
          <InfoPageView page={page} match={this.props.match} />
        </Fragment>
      )
    }
    return <Loader />
  }
}

const mapStateToProps = (state, ownProps) => {
  const slugName = `${NAMESPACE}-${ownProps.match.params.name}`
  return {
    isFetching: state.page.isFetching,
    page: state.page[slugName],
  }
}

export { InfoPage }
export default connect<*, *, *, *, *, *>(mapStateToProps, { fetchInfoPage })(
  InfoPage,
)
