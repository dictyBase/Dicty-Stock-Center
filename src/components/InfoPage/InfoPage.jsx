import React, { Component } from "react"
import { connect } from "react-redux"
import Skeleton from "react-loading-skeleton"
import InfoPageView from "./InfoPageView"
import { fetchInfoPage } from "actions/page"
import { Flex, Box } from "rebass"
import { NAMESPACE } from "constants/index"

class InfoPage extends Component {
  // set defaultprops to prevent console warnings
  static defaultProps = {
    page: {
      data: {
        attributes: {}
      }
    }
  }
  componentDidMount() {
    const { match, fetchInfoPage } = this.props
    const slugName = `${NAMESPACE}-${match.params.name}`
    fetchInfoPage(slugName)
  }
  render() {
    const { isFetching, page } = this.props

    if (!isFetching && page.data.attributes.content) {
      return <InfoPageView page={page} match={this.props.match} />
    }
    return (
      <Flex justify="center">
        <Box w={"80%"}>
          <h1>
            <Skeleton />
          </h1>
          <Skeleton count={10} />
          <br />
          <br />
          <Skeleton count={10} />
          <br />
          <br />
          <Skeleton count={10} />
        </Box>
      </Flex>
    )
  }
}

const mapStateToProps = (state, ownProps) => {
  const slugName = `${NAMESPACE}-${ownProps.match.params.name}`
  return {
    isFetching: state.page.isFetching,
    page: state.page[slugName]
  }
}

export default connect(mapStateToProps, { fetchInfoPage })(InfoPage)
