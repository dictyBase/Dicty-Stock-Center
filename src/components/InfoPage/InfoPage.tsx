import React from "react"
import { useQuery } from "@apollo/react-hooks"
import { useParams } from "react-router-dom"
import { Helmet } from "react-helmet"
import Loader from "components/common/Loader"
import InfoPageView from "./InfoPageView"
import { GET_CONTENT_BY_SLUG } from "graphql/queries"
import { NAMESPACE } from "constants/dsctypes"

const metaContent =
  "The Dicty Stock Center is a rapidly growing central repository for Dictyostelium discoideum strains and those of related species, plasmids, commonly used food bacteria, and other materials such as antibodies."
const capitalizedName = (name: string) =>
  name.charAt(0).toUpperCase() + name.slice(1)

/**
 * Fetches the data for the desired info page
 */

const InfoPage = () => {
  const { name } = useParams()
  const { loading, error, data } = useQuery(GET_CONTENT_BY_SLUG, {
    variables: {
      slug: `${NAMESPACE}-${name}`,
    },
  })

  if (loading) {
    return <Loader />
  }

  if (error || name === undefined) {
    return <div>Error fetching page content</div>
  }

  return (
    <>
      <Helmet>
        <title>{capitalizedName(name)} Information - Dicty Stock Center</title>
        <meta name="description" content={metaContent} />
      </Helmet>
      <InfoPageView data={data.contentBySlug} />
    </>
  )
}

export default InfoPage
