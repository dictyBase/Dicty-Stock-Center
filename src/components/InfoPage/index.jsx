import React, { Component } from 'react'
import { connect } from 'react-redux'
import Skeleton from 'react-loading-skeleton'
import InfoPageView from './InfoPageView'
import { fetchInfoPage } from 'actions/page'
import { Flex, Box } from 'rebass'

class InfoPage extends Component {
  displayName = 'toolbar with entity controls'
  componentDidMount() {
    const { match, fetchInfoPage } = this.props
    const slugName = 'dsc-' + match.params.name
    fetchInfoPage(slugName)
  }
  // componentWillReceiveProps(nextProps) {
  //   if (this.props.page.page !== nextProps.page.page) {
  //     const { match, fetchInfoPage } = nextProps
  //     fetchInfoPage(match.params.name)
  //   }
  // }
  render() {
    const { isFetching, content } = this.props
    if (!isFetching && content) {
      return <InfoPageView page={this.props.page} match={this.props.match} />
    }
    return (
      <Flex justify="center">
        <Box w={'80%'}>
          <h1>{this.props.title || <Skeleton />}</h1>
          <Skeleton count={10} />
          <br />
          <br />
          <Skeleton count={10} />
          <br />
          <br />
          <Skeleton count={10} />
        </Box>
      </Flex>
    )
  }
}

const mapStateToProps = state => {
  return {
    isFetching: state.page.isFetching,
    content: state.page.content,
    page: state.page
  }
}

const mapDispatchToProps = dispatch => {
  return {
    fetchInfoPage: id => {
      dispatch(fetchInfoPage(id))
    }
  }
}

export default connect(mapStateToProps, mapDispatchToProps)(InfoPage)
