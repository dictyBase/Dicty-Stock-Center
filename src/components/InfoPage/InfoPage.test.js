import React from "react"
import { shallow, mount } from "enzyme"
import sinon from "sinon"
import { InfoPage } from "./InfoPage"
import InfoPageView from "./InfoPageView"
import Loader from "components/common/Loader"

describe("InfoPage/InfoPage", () => {
  let props
  let mountedInfoPage
  const infoPage = () => {
    if (!mountedInfoPage) {
      mountedInfoPage = mount(<InfoPage {...props} />)
    }
    return mountedInfoPage
  }

  beforeEach(() => {
    props = {
      page: undefined,
      match: undefined,
      fetchInfoPage: undefined,
      isFetching: undefined,
    }
    mountedInfoPage = undefined
  })

  describe("initial render", () => {
    beforeEach(() => {
      props = {
        page: {
          data: {
            attributes: {
              content: "page content",
              name: "order",
            },
          },
        },
        match: {
          params: {
            name: "order",
          },
        },
        fetchInfoPage: () => {},
        isFetching: true,
      }
    })

    it("always renders Loader", () => {
      expect(infoPage().find(Loader).length).toBe(1)
    })
    it("calls componentDidMount", () => {
      sinon.spy(InfoPage.prototype, "componentDidMount")
      infoPage()
      expect(InfoPage.prototype.componentDidMount.calledOnce).toEqual(true)
    })
  })

  describe("after content is fetched", () => {
    beforeEach(() => {
      props = {
        page: {
          data: {
            attributes: {
              content: "page content",
              name: "order",
            },
          },
        },
        match: {
          params: {
            name: "order",
          },
        },
        fetchInfoPage: () => {},
        isFetching: false,
      }
    })

    it("renders InfoPageView", () => {
      const wrapper = shallow(<InfoPage {...props} />)
      expect(wrapper.find(InfoPageView).length).toBe(1)
    })

    it("no longer renders Loader", () => {
      const wrapper = shallow(<InfoPage {...props} />)
      expect(wrapper.find(Loader).length).toBe(0)
    })
  })
})
