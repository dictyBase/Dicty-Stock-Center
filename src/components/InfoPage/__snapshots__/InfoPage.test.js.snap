// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matching a snapshot of InfoPage/InfoPage 1`] = `
.c0 {
  padding-right: 15px;
  padding-left: 15px;
  margin-right: auto;
  margin-left: auto;
}

@media (min-width:768px) {
  .c0 {
    width: 750px;
  }
}

@media (min-width:992px) {
  .c0 {
    width: 970px;
  }
}

@media (min-width:1200px) {
  .c0 {
    width: 1170px;
  }
}

<div
  className="c0"
>
  <div
    className="sc-htpNat iZBBQa sc-bdVaJa iHZvIS"
    is={null}
  >
    <div
      className="gRrvFh"
      is={null}
    >
      <div
        className="react-loading-theme-2"
      >
        <span>
          <span
            className="react-loading-skeleton"
            style={
              Object {
                "animation": "progress 1.2s ease-in-out infinite",
              }
            }
          >
            ‌
          </span>
          <span
            className="react-loading-skeleton"
            style={
              Object {
                "animation": "progress 1.2s ease-in-out infinite",
              }
            }
          >
            ‌
          </span>
          <span
            className="react-loading-skeleton"
            style={
              Object {
                "animation": "progress 1.2s ease-in-out infinite",
              }
            }
          >
            ‌
          </span>
          <span
            className="react-loading-skeleton"
            style={
              Object {
                "animation": "progress 1.2s ease-in-out infinite",
              }
            }
          >
            ‌
          </span>
          <span
            className="react-loading-skeleton"
            style={
              Object {
                "animation": "progress 1.2s ease-in-out infinite",
              }
            }
          >
            ‌
          </span>
          <span
            className="react-loading-skeleton"
            style={
              Object {
                "animation": "progress 1.2s ease-in-out infinite",
              }
            }
          >
            ‌
          </span>
          <span
            className="react-loading-skeleton"
            style={
              Object {
                "animation": "progress 1.2s ease-in-out infinite",
              }
            }
          >
            ‌
          </span>
          <span
            className="react-loading-skeleton"
            style={
              Object {
                "animation": "progress 1.2s ease-in-out infinite",
              }
            }
          >
            ‌
          </span>
          <span
            className="react-loading-skeleton"
            style={
              Object {
                "animation": "progress 1.2s ease-in-out infinite",
              }
            }
          >
            ‌
          </span>
          <span
            className="react-loading-skeleton"
            style={
              Object {
                "animation": "progress 1.2s ease-in-out infinite",
              }
            }
          >
            ‌
          </span>
        </span>
        <br />
        <br />
        <span>
          <span
            className="react-loading-skeleton"
            style={
              Object {
                "animation": "progress 1.2s ease-in-out infinite",
              }
            }
          >
            ‌
          </span>
          <span
            className="react-loading-skeleton"
            style={
              Object {
                "animation": "progress 1.2s ease-in-out infinite",
              }
            }
          >
            ‌
          </span>
          <span
            className="react-loading-skeleton"
            style={
              Object {
                "animation": "progress 1.2s ease-in-out infinite",
              }
            }
          >
            ‌
          </span>
          <span
            className="react-loading-skeleton"
            style={
              Object {
                "animation": "progress 1.2s ease-in-out infinite",
              }
            }
          >
            ‌
          </span>
          <span
            className="react-loading-skeleton"
            style={
              Object {
                "animation": "progress 1.2s ease-in-out infinite",
              }
            }
          >
            ‌
          </span>
          <span
            className="react-loading-skeleton"
            style={
              Object {
                "animation": "progress 1.2s ease-in-out infinite",
              }
            }
          >
            ‌
          </span>
          <span
            className="react-loading-skeleton"
            style={
              Object {
                "animation": "progress 1.2s ease-in-out infinite",
              }
            }
          >
            ‌
          </span>
          <span
            className="react-loading-skeleton"
            style={
              Object {
                "animation": "progress 1.2s ease-in-out infinite",
              }
            }
          >
            ‌
          </span>
          <span
            className="react-loading-skeleton"
            style={
              Object {
                "animation": "progress 1.2s ease-in-out infinite",
              }
            }
          >
            ‌
          </span>
          <span
            className="react-loading-skeleton"
            style={
              Object {
                "animation": "progress 1.2s ease-in-out infinite",
              }
            }
          >
            ‌
          </span>
        </span>
        <br />
        <br />
        <span>
          <span
            className="react-loading-skeleton"
            style={
              Object {
                "animation": "progress 1.2s ease-in-out infinite",
              }
            }
          >
            ‌
          </span>
          <span
            className="react-loading-skeleton"
            style={
              Object {
                "animation": "progress 1.2s ease-in-out infinite",
              }
            }
          >
            ‌
          </span>
          <span
            className="react-loading-skeleton"
            style={
              Object {
                "animation": "progress 1.2s ease-in-out infinite",
              }
            }
          >
            ‌
          </span>
          <span
            className="react-loading-skeleton"
            style={
              Object {
                "animation": "progress 1.2s ease-in-out infinite",
              }
            }
          >
            ‌
          </span>
          <span
            className="react-loading-skeleton"
            style={
              Object {
                "animation": "progress 1.2s ease-in-out infinite",
              }
            }
          >
            ‌
          </span>
          <span
            className="react-loading-skeleton"
            style={
              Object {
                "animation": "progress 1.2s ease-in-out infinite",
              }
            }
          >
            ‌
          </span>
          <span
            className="react-loading-skeleton"
            style={
              Object {
                "animation": "progress 1.2s ease-in-out infinite",
              }
            }
          >
            ‌
          </span>
          <span
            className="react-loading-skeleton"
            style={
              Object {
                "animation": "progress 1.2s ease-in-out infinite",
              }
            }
          >
            ‌
          </span>
          <span
            className="react-loading-skeleton"
            style={
              Object {
                "animation": "progress 1.2s ease-in-out infinite",
              }
            }
          >
            ‌
          </span>
          <span
            className="react-loading-skeleton"
            style={
              Object {
                "animation": "progress 1.2s ease-in-out infinite",
              }
            }
          >
            ‌
          </span>
        </span>
        <br />
        <br />
        <span>
          <span
            className="react-loading-skeleton"
            style={
              Object {
                "animation": "progress 1.2s ease-in-out infinite",
              }
            }
          >
            ‌
          </span>
          <span
            className="react-loading-skeleton"
            style={
              Object {
                "animation": "progress 1.2s ease-in-out infinite",
              }
            }
          >
            ‌
          </span>
          <span
            className="react-loading-skeleton"
            style={
              Object {
                "animation": "progress 1.2s ease-in-out infinite",
              }
            }
          >
            ‌
          </span>
          <span
            className="react-loading-skeleton"
            style={
              Object {
                "animation": "progress 1.2s ease-in-out infinite",
              }
            }
          >
            ‌
          </span>
          <span
            className="react-loading-skeleton"
            style={
              Object {
                "animation": "progress 1.2s ease-in-out infinite",
              }
            }
          >
            ‌
          </span>
        </span>
      </div>
    </div>
  </div>
</div>
`;
